// appium-ios-simulator/build/lib/extensions/keychain.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const lodash_1 = __importDefault(require("lodash"));
const path_1 = __importDefault(require("path"));
const support_1 = require("@appium/support");
const logger_1 = __importDefault(require("../logger"));
const teen_process_1 = require("teen_process");
const utils_1 = require("../utils");
const extensions = {};
/**
 * Resolve full path to Simlator's LaunchDaemons root folder
 *
 * @returns {Promise<string>} Full path to Simlator's LaunchDaemons root folder
 */
extensions.getLaunchDaemonsRoot = async function getLaunchDaemonsRoot() {
    const devRoot = await (0, utils_1.getDeveloperRoot)();
    return path_1.default.resolve(devRoot, 'Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/System/Library/LaunchDaemons');
};
/**
 * Create the backup of keychains folder.
 * The previously created backup will be automatically
 * deleted if this method was called twice in a row without
 * `restoreKeychains` being invoked.
 *
 * @this {import('../simulator-xcode-14').default}
 * @returns {Promise<boolean>} True if the backup operation was successfull.
 */
extensions.backupKeychains = async function backupKeychains() {
    if (!await support_1.fs.exists(this.keychainPath)) {
        return false;
    }
    const backupPath = await support_1.tempDir.path({
        prefix: `keychains_backup_${Math.floor((1 + Math.random()) * 0x10000).toString(16).substring(1)}`,
        suffix: '.zip',
    });
    const zipArgs = [
        '-r', backupPath,
        `${this.keychainPath}${path_1.default.sep}`
    ];
    logger_1.default.debug(`Creating keychains backup with 'zip ${zipArgs.join(' ')}' command`);
    await (0, teen_process_1.exec)('zip', zipArgs);
    // @ts-ignore
    if (lodash_1.default.isString(this._keychainsBackupPath) && await support_1.fs.exists(this._keychainsBackupPath)) {
        // @ts-ignore
        await support_1.fs.unlink(this._keychainsBackupPath);
    }
    // @ts-ignore
    this._keychainsBackupPath = backupPath;
    return true;
};
/**
 * Restore the previsouly created keychains backup.
 *
 * @param {string[]} excludePatterns - The list
 * of file name patterns to be excluded from restore. The format
 * of each item should be the same as '-x' option format for
 * 'unzip' utility. This can also be a comma-separated string,
 * which is going be transformed into a list automatically,
 * for example: '*.db*,blabla.sqlite'
 * @this {import('../simulator-xcode-14').default}
 * @returns {Promise<boolean>} If the restore opration was successful.
 * @throws {Error} If there is no keychains backup available for restore.
 */
extensions.restoreKeychains = async function restoreKeychains(excludePatterns = []) {
    if (!lodash_1.default.isString(this._keychainsBackupPath) || !await support_1.fs.exists(this._keychainsBackupPath)) {
        throw new Error(`The keychains backup archive does not exist. ` +
            `Are you sure it was created before?`);
    }
    if (lodash_1.default.isString(excludePatterns)) {
        excludePatterns = excludePatterns.split(',').map((x) => x.trim());
    }
    const isServerRunning = await this.isRunning();
    let plistPath;
    if (isServerRunning) {
        plistPath = path_1.default.resolve(await this.getLaunchDaemonsRoot(), 'com.apple.securityd.plist');
        if (!await support_1.fs.exists(plistPath)) {
            throw new Error(`Cannot clear keychains because '${plistPath}' does not exist`);
        }
        await this.simctl.spawnProcess(['launchctl', 'unload', plistPath]);
    }
    try {
        await support_1.fs.rimraf(this.keychainPath);
        await (0, support_1.mkdirp)(this.keychainPath);
        const unzipArgs = [
            '-o', this._keychainsBackupPath,
            ...(lodash_1.default.flatMap(excludePatterns.map((x) => ['-x', x]))),
            '-d', '/'
        ];
        logger_1.default.debug(`Restoring keychains with 'unzip ${unzipArgs.join(' ')}' command`);
        await (0, teen_process_1.exec)('unzip', unzipArgs);
        await support_1.fs.unlink(this._keychainsBackupPath);
        this._keychainsBackupPath = null;
    }
    finally {
        if (isServerRunning && plistPath) {
            await this.simctl.spawnProcess(['launchctl', 'load', plistPath]);
        }
    }
    return true;
};
/**
 * Clears Keychains for the particular simulator in runtime (there is no need to stop it).
 *
 * @this {import('../simulator-xcode-14').default}
 * @throws {Error} If keychain cleanup has failed.
 */
extensions.clearKeychains = async function clearKeychains() {
    const plistPath = path_1.default.resolve(await this.getLaunchDaemonsRoot(), 'com.apple.securityd.plist');
    if (!await support_1.fs.exists(plistPath)) {
        throw new Error(`Cannot clear keychains because '${plistPath}' does not exist`);
    }
    await this.simctl.spawnProcess(['launchctl', 'unload', plistPath]);
    try {
        if (await support_1.fs.exists(this.keychainPath)) {
            await support_1.fs.rimraf(this.keychainPath);
            await (0, support_1.mkdirp)(this.keychainPath);
        }
    }
    finally {
        await this.simctl.spawnProcess(['launchctl', 'load', plistPath]);
    }
};
exports.default = extensions;
